streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java

```diff
@@ -887,10 +887,30 @@ private void checkIfUnexpectedUserSpecifiedConsumerConfig(final Map<String, Obje
         // consumer/producer configurations, log a warning and remove the user defined value from the Map.
         // Thus the default values for these consumer/producer configurations that are suitable for
         // Streams will be used instead.
-        final Object maxInFlightRequests = clientProvidedProps.get(ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION);
-        if (eosEnabled && maxInFlightRequests != null && 5 < (int) maxInFlightRequests) {
-            throw new ConfigException(ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION + " can't exceed 5 when using the idempotent producer");
+
+        if (eosEnabled) {
+            final Object maxInFlightRequests = clientProvidedProps.get(ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION);
+
+            if (maxInFlightRequests != null) {
+                final int maxInFlightRequestsAsInteger;
+                if (maxInFlightRequests instanceof String) {
+                    try {
+                        maxInFlightRequestsAsInteger = Integer.parseInt((String) maxInFlightRequests);
+                    } catch (final NumberFormatException e) {
+                        throw new ConfigException("Config value for " + ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION + " could not be parsed into an integer");
+                    }
+                } else if (maxInFlightRequests instanceof Number) {
+                    maxInFlightRequestsAsInteger = ((Number) maxInFlightRequests).intValue();
+                } else {
+                    throw new ConfigException("Config value (with type " + maxInFlightRequests.getClass().getName() + " for " + ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION + " could not be converted into an int.");
+                }
+
+                if (5 < maxInFlightRequestsAsInteger) {
```

Comment:
super nit: maybe reverse statement  to `maxInFlightRequestsAsInteger > 5` IMHO easier to grok, but this is highly opinionated, so feel free to ignore.